### PR TITLE
rebase -i: support --ignore-date

### PR DESCRIPTION
**Pull request by r1walz**
Let's review this one.

### CODE REVIEW

File: t/t3431-rebase-options-compatibility.sh
Comment (r1walz) [inline t/t3431-rebase-options-compatibility.sh]: Ignore test file.

File: sequencer.c
Comment (tgummerer) [inline sequencer.c]: We should check that we don't run over the end of the string here, e.g. if somebody changes the author script on disk, we still shouldn't read some random memory.

Even better I think we can replace this `while` loop with a `memchr` to get the length of the bits we need to split off, and, add that to the `strbuf` using `strbuf_add` and then `datestamp` it.

Other than being easier to read, I think this also allows us to plug a memory leak, as we are leaking the original `author` right now.  (I think we're also leaking it currently, but now we'd leak two instances)

I also wonder if it would be better to traverse this string from the end, rather than from the front.  I don't think we allow `>` in the author name, so this should work fine in all cases, but I'd still feel more comfortable going over something that isn't user controlled.

Also I think `author` could be `NULL` here right, if the `is_rebase_i` clause is not satisfied.  We should also check for that I think.

File: sequencer.c
Comment (tgummerer) [inline sequencer.c:3523]: Convenient :)

File: sequencer.c
Comment (tgummerer) [inline sequencer.c:1536]: This is the same code as above, should we factor this out into a separate function?

**Review by tgummerer — COMMENTED** [review]
Thanks!  Quite a lot to unpack in that single line there, though I think it's relatively easy to fix :)

Nice work, looking forward to reading the docs and the tests for this as well.

File: sequencer.c
Comment (r1walz) [reply sequencer.c]: AFAIK, --date sets AUTHOR TZ to UTC. But we are using local TZ in the logic. Should be changed ASAP.

File: builtin/rebase.c
Comment (newren) [inline builtin/rebase.c]: I know we're just copying git-am and earlier git-rebase here, but the name "--ignore-date" is just not descriptive enough.  In am's case it's slightly more defensible as an email message has a single date and it's known that it is used to set the author date.  But even there it feels too vague.  In the context of rebase, it's just bad -- there are clearly two dates we are working with.

We really ought to rename this to something else ("--author-date-is-committer-date"? "--ignore-author-date"?), and just keep --ignore-date around as an alias for the more precise name.

File: sequencer.c
Comment (newren) [reply sequencer.c:1536]: ...and perhaps also move the opts->committer_date_is_author_date block into that new function since it is also repeated?

**Review by newren — COMMENTED** [review]
Only two small comments other than what Thomas found.

File: sequencer.c
Comment (r1walz) [reply sequencer.c]: > Also I think author could be NULL here right, if the is_rebase_i clause is not satisfied. We should also check for that I think.

`--ignore-date` is only set by rebase -i so I think we are clear on this one.

Comment (tgummerer) [reply sequencer.c]: > `--ignore-date` is only set by rebase -i so I think we are clear on this one.

Ah yes of course, that makes sense.  It would be nice to have a `BUG()` statement then, indicating that we know about this, and helping developers in the future if they want to add another codepath here, e.g. if they want to start having a similar functionality in `cherry-pick`.

File: Documentation/git-rebase.txt
Comment (tgummerer) [inline Documentation/git-rebase.txt:391]: This is now not quite correct for am based rebases.  `--author-date-is-committer-date` doesn't exist there, and I don't see this patch implementing it.  Maybe introducing that flag in a preparatory patch would be nice.

File: Documentation/git-rebase.txt
Comment (tgummerer) [inline Documentation/git-rebase.txt]: s/comitter/committer/

File: sequencer.c
Comment (tgummerer) [inline sequencer.c]: It seems to me that `amend_author` is only used to free `author` when we get it from `get_author`, instead of from the parameter that's passed in (and should not be free'd.

So it seems to me that we should always assign `author` to `amend_author` here to avoid ano memory leaks.  At that point `amend_author` is no longer the right name, so maybe we should rename it to `author_to_free` or something in a preparatory patch?

File: t/t3433-rebase-options-compatibility.sh
Comment (tgummerer) [inline t/t3433-rebase-options-compatibility.sh:85]: That doesn't sound quite right.  The default time zone when committing should always be the timezone that's set up where the tests are run.  If we alwayse get UTC in the tests here, that's probably and indication that the code doesn't deal with timezones correctly.

And we don't set the timezone to UTC in the tests anywhere, as far as I can see.

File: sequencer.c
Comment (tgummerer) [inline sequencer.c]: Maybe expand this to something like "ignore_date can only be used with rebase -i, which must set the author befor committing the tree"?

File: t/t3433-rebase-options-compatibility.sh
Comment (r1walz) [reply t/t3433-rebase-options-compatibility.sh:85]: > The default time zone when committing should always be the timezone that's set up where the tests are run.

This is correct. But ...

> And we don't set the timezone to UTC in the tests anywhere, as far as I can see.

... we're actually setting timezone in [`test-lib.sh`](https://github.com/git/git/blob/v2.22.0/t/test-lib.sh#L368).

File: sequencer.c
Comment (r1walz) [reply sequencer.c]: Yes, it is indeed used to do the aforementioned. Let's account all your suggested changes in some commits.

File: t/t3433-rebase-options-compatibility.sh
Comment (tgummerer) [reply t/t3433-rebase-options-compatibility.sh:85]: > ... we're actually setting timezone in test-lib.sh.

Ah fair enough, I missed that. Thanks for the pointer.  I tried to `cat-file` a commit in t1010 to test this, but only now realized that we're using `test_tick` there, which confusingly sets the timezone to -0700.    So we're all good here, sorry for the noise.

File: t/t3433-rebase-options-compatibility.sh
Comment (r1walz) [inline t/t3433-rebase-options-compatibility.sh]: How about using `git show HEAD --pretty="format:%ai"` instead?

Comment (tgummerer) [reply t/t3433-rebase-options-compatibility.sh]: Yeah, I think that would be preferrable.

**Comment by r1walz** [comment]
@tgummerer @newren Let's have a final call before sending this to the ML.

File: sequencer.c
Comment (newren) [inline sequencer.c:1450]: What if someone passes both --committer-date-is-author-date and --ignore-date?  Do they expect author date and committer date to get swapped?  Right now, this appears to make --ignore-date override --committer-date-is-author-date -- perhaps we should error out instead?  (Or maybe this is fine as-is because git-am also surprisingly just makes --ignore-date silently override --committer-date-is-author-date, without documenting it, so you've at least matched the behavior elsewhere...)

Not sure you need to do anything here, especially since you do match git-am behavior, but it might be worth pointing out in your cover letter.

**Review by newren — COMMENTED** [review]
Looks good, but (as a very micro nit:) you have a double space after a comma in the commit message; might be nice to clean that up.

**Review by newren — APPROVED** [review]
Wahoo!  I like it.

**Review by tgummerer — COMMENTED** [review]
> Commit 554377d ("rebase -i: support --ignore-date", 2019-08-05)

Unfortunately you can't refer to a commit like that when it's part of a patch series, as the commit ID will be different once Junio actually applies the earlier patch.  Instead, the patch can just be referred as "a previous patch" or something like that.

I would also make what you have as the second commit in the series ("sequencer: rename amend_author to author_to_free") the first commit, as it's really a preparatory step that makes the commit after it easier to read.  Right now it's more of an "oops, we should have named this differently, let's do it now" commit, which we generally prefer not to do in this project.

Other than those two nits the code looks good to me 🎉

**Comment by r1walz** [comment]
--- sent to the [Mailing List](https://public-inbox.org/git/20190806173638.17510-1-rohit.ashiwal265@gmail.com/) ---

**Pull request closed without merging**
